### DISCUSSION TITLE
How to find failing crate?

### DISCUSSION THREAD

**Discussion by pjljvandelaar** [alire-project/alire#1188] [Q&A] [answered]
Dear Alire,

I really like the Alire Crate Status page: I want to make my crate as usable as possible!
However, I noticed that regularly the build fails.
Yet not due to my crate, but due to crates that my crate uses.

For example
https://alire-crate-ci.ada.dev/status/re/rejuvenation/22.0.1/linux/fedora-33/gnat=12.1.2/rejuvenation-22.0.1.yaml.log
shows
```
  ...
   [Ada]          unicode-names-bopomofo_extended.ads
   [Ada]          unicode-names-tagalog.ads
   [Ada]          unicode-names-latin_extended_d.ads
   [Ada]          unicode-names-sup_symbols_pictographs.ads
   [Ada]          unicode-names-specials.ads
   [Ada]          unicode-names-combining_diacritical_sup.ads
   [C]            gpr_imports.c
   [Ada]          gpr-debug.adb
   [Ada]          gpr-pp.adb
   [Ada]          gpr-compilation.adb
   [Ada]          gpr-sinput.adb
   [Ada]          gpr-opt.ads
   [Ada]          gpr-compilation-process-waiter.adb
   [Ada]          gpr-strt.adb
   [Ada]          gpr-script.adb
   [Ada]          gpr-compilation-slave.adb
   [Ada]          gpr-names.adb
   [Ada]          gpr-knowledge.adb
sax-pointers.ads:57:58: missing ";"
sax-pointers.ads:62:46: "Encapsulated" is undefined (more references follow)
sax-utils.ads:171:53: "Smart_Pointers" is not the name of a generic package
sax-utils.ads:174:28: "Symbol_Table_Pointers" is undefined (more references follow)
schema-validators.ads:912:44: "Smart_Pointers" is not the name of a generic package
schema-validators.ads:914:28: "XML_Grammars" is undefined (more references follow)
schema-validators.ads:914:28: possible misspelling of "XML_Grammar"
```

I have a number of questions:
* Does alire contact/notify crate owners informing them that their crate fails to build correctly?
* Does alire maintain a list of incorrect crates (for a given platform) and use this to NOT build crates that depend on such a crate on that platform?
* Are there issue templates to report failing builds to their crate owners?
* How do I know which crate(s) contain these failing files? 

Thanks in advance for your answers

**Comment by mosteo** [comment, marked as answer]
> * Does alire contact/notify crate owners informing them that their crate fails to build correctly?

Not as a matter of course.

> * Does alire maintain a list of incorrect crates (for a given platform) and use this to NOT build crates that depend on such a crate on that platform?

No. Even if a crate is failing due to bad dependencies, it is still a crate that isn't readily usable and we want to know about it.

> * Are there issue templates to report failing builds to their crate owners?

No. I'd go directly to the crate issues page (if hosted on github ofc).

> * How do I know which crate(s) contain these failing files?

In general, I'd say the build logs should make easy to pinpoint the offending crate. I'm not sure if there's a way using `gprbuild` to locate the project file that brings in a file into the build. This could serve to have something like `alire show --owner-crate file.adb`.

**Reply by pjljvandelaar** [reply]
I would really value something like `alire show --owner-crate file.adb`

**Reply by mosteo** [reply]
On second thought I'm not sure if this adds any value over a regular filesystem `find`? We wouldn't be able to use gprbuild on a remote crate and once you have it on disk with dependencies its moot. (Besides, my cursory look doesn't seem to show any such capability from gprbuild.)

**Reply by pjljvandelaar** [reply]
Indeed `find` does the trick (since dependencies directories have the package name nicely encode in them).
Yet, on windows I always have to struggle to find the equivalent of `find`.
